lib/FrontendTool/FrontendTool.cpp

```diff
@@ -135,6 +135,93 @@ swift::frontend::utils::escapeForMake(StringRef raw,
+// an auxiliary target as its first argument, and the main target on which
+// it should depend as its second argument.
+//
+// The final structure of the dependency file is therefore:
+//
+// /Path/To/Most/Preferred/Output/File.o:
+//   <Big Ol' List Of Files...>
+// /Path/To/Supplementary/Output/File.swiftdoc:
+//   /Path/To/Most/Preferred/Output/File.o
+// /Path/To/Supplementary/Output/File.swiftsourceinfo:
+//   /Path/To/Most/Preferred/Output/File.o
+// ...
+static void forEachMinimalMakeStyleTarget(
+    const FrontendOptions &opts,
+    const InputFile &input,
+    llvm::function_ref<void(StringRef)> mainTarget,
+    llvm::function_ref<void(StringRef, StringRef)> remainingTarget) {
+  llvm::SmallVector<const std::string *, 8> preferredOutputs;
+  llvm::StringSet<> seenTargets;
+
+  // Try to compute the main file from the action we're given.  If we're just
+  // emitting a module interface or running merge-modules, we'll pick up the
+  // module output path first from the list below.
+  if (opts.RequestedAction != FrontendOptions::ActionType::EmitModuleOnly &&
+      opts.RequestedAction != FrontendOptions::ActionType::MergeModules) {
+    if (opts.InputsAndOutputs.isWholeModule()) {
+      opts.InputsAndOutputs.forEachInputProducingAMainOutputFile([&](const InputFile &input) {
+        // Unique the file names we're pushing as targets. Because multiple
+        // source files are going to map to the same main output file.
+        if (seenTargets.insert(input.outputFilenameRef()).second) {
+            preferredOutputs.push_back(&input.outputFilenameRef());
+        }
+        // Don't early-exit.
+        return false;
+      });
+    } else {
+      preferredOutputs.push_back(&input.outputFilenameRef());
+    }
+  }
+
+  // Now append the remaining output paths in order of preference for their
+  // being the main target given no other choice.
+  const SupplementaryOutputPaths &outs =
+      input.getPrimarySpecificPaths().SupplementaryOutputs;
+  preferredOutputs.append({
+    &outs.ModuleOutputPath,
+    &outs.ModuleDocOutputPath,
+    &outs.ModuleInterfaceOutputPath,
+    &outs.ObjCHeaderOutputPath,
+    &outs.ModuleSourceInfoOutputPath,
+  });
+
+  // Slice the collection at the first non-empty candidate target.  This will
+  // be the target all the others depend on.
+  auto it = llvm::find_if(preferredOutputs, [](const std::string *str) {
+    return !str->empty();
+  });
+
+  // No targets?  There's nothing to do here.
+  if (it == std::end(preferredOutputs))
+    return;
+
+  // Apply the function to the most preferred target.
+  const std::string *const firstNonEmptyTarget(*it);
+  mainTarget(*firstNonEmptyTarget);
+
+  // Now scan forward to apply the supplementary function
+  // to the remaining targets.
+  for (++it; it != std::end(preferredOutputs); ++it) {
+    if (!(*it)->empty()) {
+      remainingTarget(**it, *firstNonEmptyTarget);
+    }
+  }
+}
```

Comment:
This function takes up almost the whole of my screen height (40") without the comment. Could it be broken up into about three pieces? Your excellent comments illustrate that each piece does have a cogent separable thought to it.